### PR TITLE
Go refactoring

### LINKED ISSUE #219: Moving Programming language file to Algorithms folder
- [x] C#
- [x] JavaScript
- [x] Python
- [x] Go
- [x] Java
- [x] C
- [x] C++
- [x] Rust
- [x] Haskell

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Review by ming-tsai — APPROVED** [review]
Hi @atin,
Thanks and nice work 🎉🎊🥇

**Comment by ming-tsai** [comment]
Sorry, some recommendations
- It is possible to add a GitHub action for build all Go file?
- It is possible to check if has Linter for Go file?

**Review by ming-tsai — CHANGES_REQUESTED** [review]
Sorry, some recommendations
- It is possible to add a GitHub action for build all Go file?
- It is possible to check if has Linter for Go file?

**Comment by atin** [comment]
Hey @ming-tsai @cclauss Any idea why `golangci-lint` is failing? Also is it possible to test the actions without pushing the changes?

**Comment by ming-tsai** [comment]
Let me check, but could you rename the file's name to `goland`?

**Comment by ming-tsai** [comment]
Did you test on the local?

**Comment by atin** [comment]
> Let me check, but could you rename the file's name to `goland`?

You mean changing the folder name from `Go` to `Golang`? I think `Go` would be better since that is the official name of the language I think.

> Did you test on the local?

What do you mean? Not sure how to test on local.

File: .github/workflows/go_lint_and_test.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+name: golangci-lint
```
Comment (ming-tsai) [inline .github/workflows/go_lint_and_test.yml:1]: Please rename this file's name to `go`, and the name attribute change `Go`

File: algorithms/Go/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Python
```
Comment (ming-tsai) [inline algorithms/Go/README.md:1]: Here use the programming language correctly

File: .github/workflows/go_lint_and_test.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+on: [push, pull_request]
+#  push:
+#    branches: [ main ]
+#    paths: '**.go'
```
Comment (ming-tsai) [inline .github/workflows/go_lint_and_test.yml:5]: Should be `**/**.go`

**Review by ming-tsai — CHANGES_REQUESTED** [review]
I added some comments

File: .github/workflows/go_lint_and_test.yml (hunk 3 above)
Comment (cclauss) [reply .github/workflows/go_lint_and_test.yml:5]: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths agrees with the current code.

Comment (ming-tsai) [reply .github/workflows/go_lint_and_test.yml:5]: Ok

**Comment by cclauss** [comment]
Please remove `-mod=readonly` from `go.yml`.

**Review by ming-tsai — DISMISSED** [review]

**Review by ming-tsai — APPROVED** [review]
I had changed by myself

**Review by ming-tsai — CHANGES_REQUESTED** [review]
Hi @atin,
Could you help me to check why lint is failling?

File: .github/workflows/go.yml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-name: golangci-lint
+name: Go
```
Comment (cclauss) [inline .github/workflows/go.yml:1]: Change line 10 to also be `Go:`

**Review by cclauss — DISMISSED** [review]
LGTM

File: .github/workflows/go.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+#    paths: '**.go'
+#  pull_request:
+#    branches: [ main ]
+#    paths: '**.go'
+jobs:
+  golangci:
+    name: lint
+    runs-on: ubuntu-latest
+    steps:
+      - run: go list -mod=readonly -e -json -compiled=true -test=true -export=true -deps=true -find=false -- ./...
```
Comment (cclauss) [inline .github/workflows/go.yml:14]: Please remove `-mod=readonly` so our tests pass.

File: .github/workflows/go.yml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+name: Go
+on: [push, pull_request]
+#  push:
+#    branches: [ main ]
+#    paths: '**.go'
+#  pull_request:
+#    branches: [ main ]
+#    paths: '**.go'
+jobs:
+  golangci:
```
Comment (cclauss) [inline .github/workflows/go.yml:10]:
```suggestion
  Go:
```

File: .github/workflows/go.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+on: [push, pull_request]
+#  push:
+#    branches: [ main ]
+#    paths: '**.go'
+#  pull_request:
+#    branches: [ main ]
+#    paths: '**.go'
+jobs:
+  Go:
+    name: lint
```
Comment (cclauss) [inline .github/workflows/go.yml:11]:
```suggestion
    name: Go
```
This does both linting and testing.

File: .github/workflows/go.yml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - run: go list -e -json -compiled=true -test=true -export=true -deps=true -find=false -- ./...
+      - uses: actions/checkout@v2
```
Comment (cclauss) [inline .github/workflows/go.yml:15]:
```suggestion
      - uses: actions/checkout@v2
      - run: go list -e -json -compiled=true -test=true -export=true -deps=true -find=false -- ./...
```
Let's checkout before we list

File: .github/workflows/go.yml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+        with:
+          version: latest
+          
+          args: ./...
+          # Optional: working directory, useful for monorepos
+          # working-directory: algorithms/Go
+          
+          # Optional: show only new issues if it's a pull request. The default value is `false`.
+          # only-new-issues: true
+      - run: go test $(go list ./... | grep -v /sbom | grep -v /rsa)
```
Comment (cclauss) [inline .github/workflows/go.yml:27]:
```suggestion
      - run: go test $(go list ./...)
```
This repo does not have directories `/sbom` and `/rsa`.

File: .github/workflows/go.yml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - run: go list -e -json -compiled=true -test=true -export=true -deps=true -find=false -- ./...
+      - uses: actions/checkout@v2
+      - name: golangci-lint
+        uses: golangci/golangci-lint-action@v2
+        with:
+          version: latest
+          
+          args: ./...
```
Comment (cclauss) [inline .github/workflows/go.yml:21]:
```suggestion
          # args: ./...
```
This seems ___optional___ according to https://github.com/golangci/golangci-lint-action

**Review by cclauss — CHANGES_REQUESTED** [review]

**Review by ming-tsai — APPROVED** [review]
Nice work

**Comment by cclauss** [comment]
Not a good idea to merge pull requests with failing tests.

**Pull request merged**